espnet/nets/pytorch_backend/e2e_transformer.py

```diff
@@ -0,0 +1,734 @@
+
+
+class MultiSequential(torch.nn.Sequential):
+    def forward(self, *args):
+        for m in self:
+            args = m(*args)
+        return args
+
+
+def repeat(N, fn):
+    """repeat module N times
+
+    :param int N: repeat time
+    :param function fn: function to generate module
+    :return: repeated modules
+    :rtype: MultiSequential
+    """
+    return MultiSequential(*[fn() for _ in range(N)])
+
+
+class EncoderLayer(nn.Module):
+    def __init__(self, size, self_attn, feed_forward, dropout):
+        super(EncoderLayer, self).__init__()
+        self.self_attn = self_attn
+        self.feed_forward = feed_forward
+        self.norm1 = LayerNorm(size)
+        self.norm2 = LayerNorm(size)
+        self.dropout = nn.Dropout(dropout)
+        self.size = size
+
+    def forward(self, x, mask):
+        """Compute encoded features
+
+        :param torch.Tensor x: encoded source features (batch, max_time_in, size)
+        :param torch.Tensor mask: mask for x (batch, max_time_in)
+        """
+        nx = self.norm1(x)
+        x = x + self.dropout(self.self_attn(nx, nx, nx, mask))
+        nx = self.norm2(x)
+        return x + self.dropout(self.feed_forward(nx)), mask
+
+
+class DecoderLayer(nn.Module):
+    def __init__(self, size, self_attn, src_attn, feed_forward, dropout):
+        super(DecoderLayer, self).__init__()
+        self.size = size
+        self.self_attn = self_attn
+        self.src_attn = src_attn
+        self.feed_forward = feed_forward
+        self.norm1 = LayerNorm(size)
+        self.norm2 = LayerNorm(size)
+        self.norm3 = LayerNorm(size)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, tgt, tgt_mask, memory, memory_mask):
+        """Compute decoded features
+
+        :param torch.Tensor tgt: decoded previous target features (batch, max_time_out, size)
+        :param torch.Tensor tgt_mask: mask for x (batch, max_time_out)
+        :param torch.Tensor memory: encoded source features (batch, max_time_in, size)
+        :param torch.Tensor memory_mask: mask for memory (batch, max_time_in)
+        """
+        x = tgt
+        nx = self.norm1(x)
+        x = x + self.dropout(self.self_attn(nx, nx, nx, tgt_mask))
+        nx = self.norm2(x)
+        x = x + self.dropout(self.src_attn(nx, memory, memory, memory_mask))
+        nx = self.norm3(x)
+        return x + self.dropout(self.feed_forward(nx)), tgt_mask, memory, memory_mask
+
+
+def subsequent_mask(size, device="cpu", dtype=torch.uint8):
+    """Create mask for subsequent steps (1, size, size)
+
+    :param int size: size of mask
+    :param str device: "cpu" or "cuda" or torch.Tensor.device
+    :param torch.dtype dtype: result dtype
+    :rtype: torch.Tensor
+    >>> subsequent_mask(3)
+    [[1, 0, 0],
+     [1, 1, 0],
+     [1, 1, 1]]
+    """
+    ret = torch.ones(size, size, device=device, dtype=dtype)
+    return torch.tril(ret, out=ret)
+
+
+class MultiHeadedAttention(nn.Module):
```

Comment:
One option is to put it in https://github.com/espnet/espnet/blob/v.0.3.0/espnet/nets/pytorch_backend/attentions.py?
You ma also reuse Tomoki's multihead attention.
Probably we need some discussions of leaving this style or split encoder/decoder/attention like RNN.
@gtache, do you have some comments?

Reply:
I think Shigeki’s implementation is better than mine.
We should use Shigeki’s one.

Reply:
I'd advise splitting as much as possible when it is not already merged, because if you decide to modularize later it will probably cause a lot of conflicts.    
The file is already 700+ lines long, and contains helper methods for the arguments and the transformer, classes for the encoder, decoder, attention, etc, so if they are not compatible with the RNN model, I'd say to put them in different files in `espnet.asr.pytorch_backend.transformer`.